### PR TITLE
BUG: read_csv not respecting object dtype when option is set

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

we are not honouring object dtype here, thoughts on performance @jbrockmendel ?

### CODE REVIEW

**Comment by phofl** [comment]
Can we get this one in?

File: pandas/io/parsers/readers.py [hunk 1]
```diff
@@ -1846,7 +1851,29 @@ def read(self, nrows: int | None = None) -> DataFrame:
             else:
                 new_rows = len(index)
 
-            df = DataFrame(col_dict, columns=columns, index=index)
+            if hasattr(self, "orig_options"):
+                dtype_arg = self.orig_options.get("dtype", None)
```
Comment (mroeschke) [inline pandas/io/parsers/readers.py:1855]: Is the `dtype` option normally applied in `_engine.read`? Just curious why it needs to be done here

Comment (phofl) [reply pandas/io/parsers/readers.py:1855]: Yes, but the DataFrame constructor infers object to string again if the option is set, which would discard the original dtype

**Comment by phofl** [comment]
cc @mroeschke gentle ping

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

File: pandas/io/parsers/readers.py [hunk 2]
```diff
@@ -1846,7 +1853,40 @@ def read(self, nrows: int | None = None) -> DataFrame:
             else:
                 new_rows = len(index)
 
-            df = DataFrame(col_dict, columns=columns, index=index)
+            if hasattr(self, "orig_options"):
```
Comment (jbrockmendel) [inline pandas/io/parsers/readers.py:1856]: can we do something more explicit than a hasattr check?

Comment (phofl) [reply pandas/io/parsers/readers.py:1856]: no, you can subclass the reader, so we don't have any control over it

Comment (jbrockmendel) [reply pandas/io/parsers/readers.py:1856]: does anybody actually do this?  i judge those people, their ethics, and their hygiene.

Comment (phofl) [reply pandas/io/parsers/readers.py:1856]: That's something I can't answer, we might want to deprecate maybe, but we are stuck with hasattr here until then

**Pull request merged**